### PR TITLE
Fix back navigation with articulation

### PR DESCRIPTION
**Pull request by mike-spa**
Resolves: #17814 

Resolves the second bug reported in the linked issue.

### CODE REVIEW

File: src/engraving/libmscore/chordrest.cpp [hunk 1]
```diff
@@ -974,7 +974,14 @@ EngravingItem* ChordRest::prevElement()
     }
     }
     staff_idx_t staffId = e->staffIdx();
-    return segment()->prevElement(staffId);
+    EngravingItem* prevItem = segment()->prevElement(staffId);
+    if (prevItem && prevItem->isNote()) {
+        Chord* prevChord = toNote(prevItem)->chord();
+        if (prevChord && prevChord->articulations().size()) {
```
Comment (Eism) [inline src/engraving/libmscore/chordrest.cpp:980]: better !empty() instead of size()

**Review by Eism — APPROVED** [review]

**Comment by mike-spa** [comment]
Actually, it works fine for articulations but not for ornaments. I need to look into it some more

**Comment by DmitryArefiev** [comment]
I tested this on Win10. It looks better then before, but there are still some issues when navigating with **Alt+LEFT arrow**
* Some **ornaments** and some **articulations** are still **skip**
* Some **notes are skip** with added ornaments 

In this score [all ornaments and articulation.zip](https://github.com/musescore/MuseScore/files/11916138/all.ornaments.and.articulation.zip)
 I marked skipped articulations/ornaments with x note and skipped notes with yellow 

https://github.com/musescore/MuseScore/assets/90187801/19499475-85dc-49ef-83ed-66b1d460865b

**Comment by mike-spa** [comment]
@DmitryArefiev I couldn't really reproduce point 2. As for point 1, as far as I could tell, it was only when passing through a barline that the articulation was skipped. That's now fixed. Let me know if it works!

**Review by oktophonie — APPROVED** [review]

**Comment by DmitryArefiev** [comment]
> @DmitryArefiev I couldn't really reproduce point 2. As for point 1, as far as I could tell, it was only when passing through a barline that the articulation was skipped. That's now fixed. Let me know if it works!

@mike-spa All good now! Thanks!



Tested #17814 on Win10, Mac13, LinuxUbuntu 22.04 - **FIXED**

File: src/engraving/libmscore/segment.cpp [hunk 2]
```diff
@@ -1677,22 +1677,18 @@ EngravingItem* Segment::prevElementOfSegment(Segment* s, EngravingItem* e, staff
...
+    std::vector<EngravingItem*>& elements = s->elist();
+    for (auto i = elements.rbegin(); i != elements.rend(); ++i) {
+        EngravingItem* item = *i;
+        if (item && item->staffIdx() == activeStaff) {
+            if (item->isChord()) {
+                auto& articulations = toChord(item)->articulations();
+                if (!articulations.empty()) {
+                    return articulations.back();
+                }
+                return toChord(item)->notes().front();
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/segment.cpp:1689]: The app will crash if this list is empty for some reason

File: src/engraving/libmscore/segment.cpp [hunk 3]
```diff
@@ -1677,22 +1677,18 @@ EngravingItem* Segment::prevElementOfSegment(Segment* s, EngravingItem* e, staff
...
 EngravingItem* Segment::lastElementOfSegment(Segment* s, staff_idx_t activeStaff)
 {
-    std::vector<EngravingItem*> elements = s->elist();
-    for (auto i = --elements.end(); i != elements.begin(); --i) {
-        if (*i && (*i)->staffIdx() == activeStaff) {
-            if ((*i)->isChord()) {
-                return toChord(*i)->notes().front();
-            } else {
-                return *i;
+    std::vector<EngravingItem*>& elements = s->elist();
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/segment.cpp:1680]: **const** std::vector<EngravingItem*>&

File: src/engraving/libmscore/segment.cpp [hunk 4]
```diff
@@ -1677,22 +1677,18 @@ EngravingItem* Segment::prevElementOfSegment(Segment* s, EngravingItem* e, staff
...
+    for (auto i = elements.rbegin(); i != elements.rend(); ++i) {
+        EngravingItem* item = *i;
+        if (item && item->staffIdx() == activeStaff) {
+            if (item->isChord()) {
+                auto& articulations = toChord(item)->articulations();
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/segment.cpp:1685]: const std::vector<Articulation*>&

File: src/engraving/libmscore/chordrest.cpp [hunk 5]
```diff
@@ -976,7 +976,14 @@ EngravingItem* ChordRest::prevElement()
     }
     }
     staff_idx_t staffId = e->staffIdx();
-    return segment()->prevElement(staffId);
+    EngravingItem* prevItem = segment()->prevElement(staffId);
+    if (prevItem && prevItem->isNote()) {
+        Chord* prevChord = toNote(prevItem)->chord();
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/chordrest.cpp:981]: const

File: src/engraving/libmscore/segment.cpp [hunk 6]
```diff
@@ -1677,22 +1677,18 @@ EngravingItem* Segment::prevElementOfSegment(Segment* s, EngravingItem* e, staff
...
 {
-    std::vector<EngravingItem*> elements = s->elist();
-    for (auto i = --elements.end(); i != elements.begin(); --i) {
-        if (*i && (*i)->staffIdx() == activeStaff) {
-            if ((*i)->isChord()) {
-                return toChord(*i)->notes().front();
-            } else {
-                return *i;
+    std::vector<EngravingItem*>& elements = s->elist();
+    for (auto i = elements.rbegin(); i != elements.rend(); ++i) {
```
Comment (RomanPudashkin) [inline src/engraving/libmscore/segment.cpp:1681]: It is better to use the name "it" when working with iterators

**Pull request merged**